lib/tether/tether_linux.go

```diff
@@ -0,0 +1,184 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tether
+
+import (
+	"fmt"
+	"io"
+	_ "net/http/pprof"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/kr/pty"
+	"github.com/vmware/vic/pkg/trace"
+)
+
+// Mkdev will hopefully get rolled into go.sys at some point
+func Mkdev(majorNumber int, minorNumber int) int {
+	return (majorNumber << 8) | (minorNumber & 0xff) | ((minorNumber & 0xfff00) << 12)
+}
+
+// childReaper is used to handle events from child processes, including child exit.
+// If running as pid=1 then this means it handles zombie process reaping for orphaned children
+// as well as direct child processes.
+func (t *tether) childReaper() {
+	signal.Notify(t.incoming, syscall.SIGCHLD)
+
+	// TODO: Call prctl with PR_SET_CHILD_SUBREAPER so that we reap regardless of pid 1 or not
+	// we already get our direct children, but not lower in the hierarchy
+
+	log.Info("Started reaping child processes")
+
+	go func() {
+		for _ = range t.incoming {
+			var status syscall.WaitStatus
+
+			func() {
+				// general resiliency
+				defer recover()
+
+				// reap until no more children to process
+				for {
+					log.Debugf("Inspecting children with status change")
+					pid, err := syscall.Wait4(-1, &status, syscall.WNOHANG, nil)
+					if pid == 0 || err == syscall.ECHILD {
+						log.Debug("No more child processes to reap")
+						break
+					}
+					if err == nil {
+						if !status.Exited() {
+							log.Debugf("Received notifcation about non-exit status change for %d:", pid)
+							// no reaping or exit handling required
+							continue
+						}
+
+						log.Debugf("Reaped process %d, return code: %d", pid, status.ExitStatus())
+
+						session, ok := t.removeChildPid(pid)
+						if ok {
+							session.ExitStatus = status.ExitStatus()
+							t.handleSessionExit(session)
+						} else {
+							// This is an adopted zombie. The Wait4 call
+							// already clean it up from the kernel
+							log.Infof("Reaped zombie process PID %d\n", pid)
+						}
+					} else {
+						log.Warnf("Wait4 got error: %v\n", err)
+					}
+				}
+			}()
+		}
+	}()
+}
+
+func (t *tether) stopReaper() {
+	defer trace.End(trace.Begin("Shutting down child reaping"))
+
+	// stop child reaping
+	log.Info("Shutting down reaper")
+	signal.Reset(syscall.SIGCHLD)
+	close(t.incoming)
+}
+
+func findExecutable(file string) error {
+	d, err := os.Stat(file)
+	if err != nil {
+		return err
+	}
+	if m := d.Mode(); !m.IsDir() && m&0111 != 0 {
+		return nil
+	}
+	return os.ErrPermission
+}
+
+// lookPath searches for an executable binary named file in the directories
+// specified by the path argument.
+// This is a direct modification of the unix os/exec core libary impl
+func lookPath(file string, env []string, dir string) (string, error) {
```

Comment:
Can we rename this please; `lookPath` is a little odd. How about `findFile` or something similar?